### ISSUE TITLE
multiplatform support

### ISSUE THREAD

**Issue by evant** [evant/kotlin-inject#1]
Migrate to a kotlin compiler plugin to support multiplatform

**Comment by evant** [comment]
Going to wait for a release of https://github.com/arrow-kt/arrow-meta as it could simplify writing a kotlin compiler plugin

**Comment by evant** [comment]
ended up going with [ksp](https://github.com/android/kotlin/tree/ksp/libraries/tools/kotlin-symbol-processing-api) instead, will get multiplatform when it does.

**Comment by JavierSegoviaCordoba** [comment]
ksp has Multiplatform support now

**Comment by evant** [comment]
@JavierSegoviaCordoba source? The issue I've been tracking is still open https://github.com/google/ksp/issues/33

**Comment by JavierSegoviaCordoba** [comment]
![image](https://user-images.githubusercontent.com/7463564/96030877-e06dbb80-0e5c-11eb-99b1-281e046bec32.png)

https://www.youtube.com/watch?v=FRCASfAMXoQ&t=1033s

**Comment by evant** [comment]
Despite what the video says this does not seem to be the case. Would recommend following the linked issue if you want updates.

**Comment by dgomolka** [comment]
ksp has this issue planned for resolve in 2021Q3🤯 that seems like a long time to wait for such a crucial (in my opinion) feature

**Comment by evant** [comment]
@dgomolka understand your frustration, but is there any action you are hoping to accomplish with your comment? Keep in mind both this project and ksp are open-source and accept contributions.

**Comment by evant** [comment]
Released v0.3.6 which has experimental kotlin js support. Note: only `js(LEGACY)` is supported by ksp for now, blocked on https://github.com/JetBrains/kotlin/pull/4264. Working on a sample as well.

**Comment by yshrsmz** [comment]
Looks like K/N is now supported 🎉 
https://github.com/google/ksp/issues/60

**Comment by evant** [comment]
Some ci stuff to work through, but have a pr up https://github.com/evant/kotlin-inject/pull/143

**Comment by evant** [comment]
Published `0.3.7-SNAPSHOT` with native support if anyone wants to try it out. Note, requires:
```
kotlin = "1.5.30"
ksp = "1.5.30-1.0.0-beta09"
```

```kotlin
repositories {
    maven { url = uri("https://oss.sonatype.org/content/repositories/snapshots") }
}
```

**Comment by saket** [comment]
I'm excited to try this out, now that KSP is stable: https://android-developers.googleblog.com/2021/09/accelerated-kotlin-build-times-with.html

**Comment by saket** [comment]
@evant was the `kotlin-inject-compiler-ksp` published correctly with support for native targets? Gradle is failing to find it:

```
* What went wrong:
Execution failed for task ':common:kspKotlinMacOS'.
> Could not resolve all files for configuration ':common:macOSCompileKlibraries'.
   > Could not resolve me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT.
     Required by:
         project :common
      > The consumer was configured to find a usage of 'kotlin-api' of a component, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native', attribute 'org.jetbrains.kotlin.native.target' with value 'macos_x64'. However we cannot choose between the following variants of me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT:20210906.175659-2:
          - javadocElements
          - sourcesElements
        All of them match the consumer attributes:
          - Variant 'javadocElements' capability me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT declares a runtime of a component:
              - Unmatched attributes:
                  - Provides documentation but the consumer didn't ask for it
                  - Provides its dependencies declared externally but the consumer didn't ask for it
                  - Provides javadocs but the consumer didn't ask for it
                  - Provides integration status but the consumer didn't ask for it
                  - Doesn't say anything about org.jetbrains.kotlin.native.target (required 'macos_x64')
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'native')
          - Variant 'sourcesElements' capability me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT declares a runtime of a component:
              - Unmatched attributes:
                  - Provides documentation but the consumer didn't ask for it
                  - Provides its dependencies declared externally but the consumer didn't ask for it
                  - Provides sources but the consumer didn't ask for it
                  - Provides integration status but the consumer didn't ask for it
                  - Doesn't say anything about org.jetbrains.kotlin.native.target (required 'macos_x64')
                  - Doesn't say anything about org.jetbrains.kotlin.platform.type (required 'native')
        The following variants were also considered but didn't match the requested attributes:
          - Variant 'apiElements' capability me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT declares an API of a component:
              - Incompatible because this component declares a component, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' and the consumer needed a component, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.native.target (required 'macos_x64')
          - Variant 'runtimeElements' capability me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT declares a runtime of a component:
              - Incompatible because this component declares a component, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'jvm' and the consumer needed a component, as well as attribute 'org.jetbrains.kotlin.platform.type' with value 'native'
              - Other compatible attribute:
                  - Doesn't say anything about org.jetbrains.kotlin.native.target (required 'macos_x64')
```

I checked the published `.module` file and couldn't find any native targets: https://oss.sonatype.org/content/repositories/snapshots/me/tatarka/inject/kotlin-inject-compiler-ksp/0.3.7-SNAPSHOT/kotlin-inject-compiler-ksp-0.3.7-20210906.175659-2.module

Btw, would this be the correct way to add the compiler plugin as a dependency?

```kotlin
val commonMain by getting {
  dependencies {
    configurations["ksp"].dependencies.add(implementation("me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT"))
    implementation("me.tatarka.inject:kotlin-inject-runtime:0.3.7-SNAPSHOT")
  }
}  
```

**Comment by evant** [comment]
Yep you have it correct. Will double check bit published properly when I get a chance, it's certainly possible it didn't.

**Comment by evant** [comment]
Note: kotlin-inject-compiler-ksp doesn't need to have native targets since it's a compiler plugin so it's only needs to be jvm. kotlin-inject-runtime is the one that needs native targets.

**Comment by evant** [comment]
Figured it out, the correct syntax for external projects is:

```kotlin
dependencies.add("ksp", "me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-SNAPSHOT")
dependencies {
    implementation("me.tatarka.inject:kotlin-inject-runtime:0.3.7-SNAPSHOT")
}
```

**Comment by evant** [comment]
Fyi updated SNAPSHOT to point to ksp stable `1.5.30-1.0.0`

**Comment by evant** [comment]
Published a [sample](https://github.com/evant/kotlin-inject-samples/tree/main/multiplatform/echo) to show how you can use kotlin-inject in your common multiplatform code. Take a particular look at https://github.com/evant/kotlin-inject-samples/blob/main/multiplatform/echo/build.gradle.kts, as it's not the most obvious how to set that up (hopefully ksp will improve this in the future).

**Comment by saket** [comment]
Thank you, this is super useful. I'm blocked on some libraries before I can update my kotlin version to 1.5.30. I'll report back once I'm able to do that!

**Comment by evant** [comment]
Released [v0.3.7-RC]() which depends on kotlin `1.6.0-RC` and ksp `1.6.0-RC-1.0.1-RC`. For what I imagine is the common case of generating components just in the common code, you'll want the following setup:

```kotlin
plugins {
    kotlin("multiplatform") version "1.6.0-RC"
    id("com.google.devtools.ksp") version "1.6.0-RC-1.0.1-RC"
}

dependencies {
    add("kspMetadata", "me.tatarka.inject:kotlin-inject-compiler-ksp:0.3.7-RC")
}

kotlin {
    // setup targets    

    sourceSets {
        commonMain {
            dependencies {
                implementation("org.jetbrains.kotlin:kotlin-stdlib-common")
                implementation("me.tatarka.inject:kotlin-inject-runtime:0.3.7-RC")
            }
        }
        // other source sets
    }
}

// Generate common code with ksp instead of per-platform, hopefully this won't be needed in the future.
// https://github.com/google/ksp/issues/567
kotlin.sourceSets.commonMain {
    kotlin.srcDir("build/generated/ksp/commonMain/kotlin")
}
tasks.withType<org.jetbrains.kotlin.gradle.dsl.KotlinCompile<*>>().all {
    if (name != "kspKotlinMetadata") {
        dependsOn("kspKotlinMetadata")
    }
}
```

you can also check out the [kotlin native sample project](https://github.com/evant/kotlin-inject-samples/tree/main/multiplatform/echo), which has been updated.

**Comment by wverdese** [comment]
I tried out what you suggested on a new KMM project, and it didn't compile. 

Here's the gradle file:
https://github.com/wverdese/kmm-kotlin-inject/blob/main/shared/build.gradle.kts

The issue:
https://gist.github.com/wverdese/3e7f2c4fe14c028c50253f0e700b8541

Any suggestion?

**Comment by evant** [comment]
@wverdese what version of gradle are you running? You may need to update it.

**Comment by wverdese** [comment]
Ah, damn good point, I was still on 7.0.3. Upgrading the gradle wrapper to 7.2 fixed it. 🎉

**Comment by evant** [comment]
Released [v0.4.0](https://github.com/evant/kotlin-inject/releases/tag/v0.4.0) which has kotln native support and depends on a stable version of kotlin (1.5.31)

**Comment by evant** [comment]
All platforms should be supported as of ksp 1.6.0-1.0.2. Closing this issue. Feel free to open a new issue if you run into any platform-specific issues.
